### PR TITLE
✨ source-stripe: use concurrent source to read multiple streams concurrently

### PR DESCRIPTION
**Pull request by girarda**
## What
* Update the stripe connector to read multiple streams concurrently

## How
* Extend ConcurrentSourceAdapter instead of AbstractSource

## 🚨 User Impact 🚨
Full refresh syncs will be faster

## Pre-merge Actions
*Expand the relevant checklist and delete the others.*

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- Connector version is set to `0.0.1`
    - `Dockerfile` has version `0.0.1`
- Documentation updated
    - Connector's `README.md`
    - Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - `docs/integrations/<source or destination>/<name>.md` including changelog with an entry for the initial version. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - `docs/integrations/README.md`

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Connector Generator</strong></summary>

- Issue acceptance criteria met
- PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/resources/pull-requests-handbook)
- If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:generateScaffolds` then checking in your changes
- Documentation which references the generator is updated as needed

</details>

<details><summary><strong>Updating the Python CDK</strong></summary>

### Airbyter

Before merging:
- Pull Request description explains what problem it is solving
- Code change is unit tested
- Build and my-py check pass
- Smoke test the change on at least one affected connector
   - On Github: Run [this workflow](https://github.com/airbytehq/airbyte/actions/workflows/connectors_tests.yml), passing `--use-local-cdk --name=source-<connector>` as options
   - Locally: `airbyte-ci connectors --use-local-cdk --name=source-<connector> test`
- PR is reviewed and approved
      
After merging:
- [Publish the CDK](https://github.com/airbytehq/airbyte/actions/workflows/publish-cdk-command-manually.yml)
   - The CDK does not follow proper semantic versioning. Choose minor if this the change has significant user impact or is a breaking change. Choose patch otherwise.
   - Write a thoughtful changelog message so we know what was updated.
- Merge the platform PR that was auto-created for updating the Connector Builder's CDK version
   - This step is optional if the change does not affect the connector builder or declarative connectors.

</details>

### CODE REVIEW

**Review by maxi297 — CHANGES_REQUESTED** [review]
see https://github.com/airbytehq/airbyte/pull/32908#issuecomment-1836202144

**Review by clnoll — COMMENTED** [review]
On the face of it the change looks fine but would like to take another look once the CAT failure is addressed.

File: airbyte-integrations/connectors/source-stripe/source_stripe/source.py [hunk 1]
```diff
@@ -76,18 +87,8 @@ def validate_and_fill_with_defaults(config: MutableMapping) -> MutableMapping:
...
-            except pendulum.parsing.exceptions.ParserError as e:
-                message = f"Invalid start date {start_date}. Please use YYYY-MM-DDTHH:MM:SSZ format."
-                raise AirbyteTracedException(
-                    message=message,
-                    internal_message=message,
-                    failure_type=FailureType.config_error,
-                ) from e
-        else:
-            start_date = pendulum.datetime(2017, 1, 25).int_timestamp
-        config["start_date"] = start_date
```
Comment (clnoll) [inline airbyte-integrations/connectors/source-stripe/source_stripe/source.py:90]: So is it the case that we never really needed to mutate the config?

**Review by clnoll — APPROVED** [review]
LGTM!

**Review by maxi297 — APPROVED** [review]
LGTM! Have we checked if we have the expected benefits? My memory is telling me that we expect CPU to always reach the limit instead of being low while processing streams without data which would lead to shorter syncs. Have confirmed this assumption using https://app.datadoghq.com/dashboard/b2x-7y7-izf/concurrent-cdk---stripe-resource-usage?

**Comment by girarda** [comment]
Good question @maxi297. The CPU is not 100% pegged because there's no real guarantee that there will always be records ready to be processed, but the CPU is able to peek:
Screenshot 2023-12-11 at 4 51 26 PM

Compared with CPU usage on 5.0.2 where the CPU usage only have short spikes above the 50% mark:
Screenshot 2023-12-11 at 4 52 07 PM

File: airbyte-integrations/connectors/source-stripe/integration_tests/expected_records.jsonl [hunk 2]
```diff
@@ -17,7 +17,7 @@
...
...m_fields": null, "description": null, "footer": null, "metadata": {}, "rendering_options": null}}, "livemode": false, "locale": null, "metadata": {}, "mode": "payment", "payment_intent": "pi_3O2XZ1EcXtiJtvvh0zWGn33E", "payment_link": null, "payment_method_collection": "always", "payment_method_configuration_details": {"id": "pmc_1MC0oMEcXtiJtvvhmhbSUwTJ", "parent": null}, "payment_method_options": {"us_bank_account": {"financial_connections": {"permissions": ["payment_method"], "prefetch": []}, "verification_method": "automatic"}, "wechat_pay": {"app_id": null, "client": "web"}}, "payment_method_types": ["card", "alipay", "klarna", "link", "us_bank_account", "wechat_pay", "cashapp"], "payment_status": "unpaid", "phone_number_collection": {"enabled": false}, "recovered_from": null, "setup_intent": null, "shipping_address_collection": null, "shipping_cost": null, "shipping_details": null, "shipping_options": [], "status": "expired", "submit_type": null, "subscription": null, "success_url": "https://example.com/success", "total_details": {"amount_discount": 0, "amount_shipping": 0, "amount_tax": 0}, "ui_mode": "hosted", "url": null, "updated": 1697627124}, "emitted_at": 1697627275062}
```
Comment (girarda) [inline airbyte-integrations/connectors/source-stripe/integration_tests/expected_records.jsonl:20]: the record's "custom_text" has a new field `"after_submit": null`

**Pull request merged**